nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -529,4 +540,397 @@ impl DataStore {
+                ))
+                .check_if_exists::<Instance>(instance_id)
+                .execute_and_check(
+                    &*self.pool_connection_authorized(opctx).await?,
+                )
+                .await
+                .map(|r| {
+                    // If we successfully updated the instance record, we have
+                    // acquired the lock; otherwise, we haven't --- either because
+                    // our generation is stale, or because the instance is already locked.
+                    let locked = match r.status {
+                        UpdateStatus::Updated => true,
+                        UpdateStatus::NotUpdatedButExists => false,
+                    };
+                    (r.found, locked)
+                })
+                .map_err(|e| {
+                    public_error_from_diesel(
+                        e,
+                        ErrorHandler::NotFoundByLookup(
+                            ResourceType::Instance,
+                            LookupType::ById(instance_id),
+                        ),
+                    )
+                })?;
+        }
+    }
+
+    /// Release the instance-updater lock acquired by
+    /// [`DataStore::instance_updater_lock`].
+    ///
+    /// This method will unlock the instance if (and only if) the lock is
+    /// currently held by the provided `saga_lock_id`. If the lock is held by a
+    /// different saga UUID, the instance will remain locked. If the instance
+    /// has already been unlocked, this method will return `false`.
+    ///
+    /// # Arguments
+    ///
+    /// - `authz_instance`: the instance to attempt to unlock
+    /// - `saga_lock_id`: the UUID of the saga that's releasing the lock on this
+    ///   instance
+    pub async fn instance_updater_unlock(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        saga_lock_id: Uuid,
+    ) -> Result<bool, Error> {
+        use db::schema::instance::dsl;
+
+        let instance_id = authz_instance.id();
+
+        let unlocked = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(authz_instance.id()))
+            // Only unlock the instance if:
+            // - the provided updater ID matches that of the saga that has
+            //   currently locked this instance.
+            .filter(dsl::updater_id.eq(Some(saga_lock_id)))
+            .set((
+                dsl::updater_gen.eq(dsl::updater_gen + 1),
+                dsl::updater_id.eq(None::<Uuid>),
+            ))
+            .check_if_exists::<Instance>(instance_id)
+            .execute_and_check(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map(|r| match r.status {
+                UpdateStatus::Updated => true,
+                // TODO(eliza): should this be an error?
+                UpdateStatus::NotUpdatedButExists => false,
```

Comment:
I think this case does indicate a programming error in Nexus. I would be sort of inclined to convert it to an error. 

IIUC, we'll hit this case whenever the instance ID appears in the instance table, but one of the filter conditions didn't hold. This implies that either that

1. someone is dropping a lock for a deleted instance, or
2. the lock ID is different from what the caller supplied

Case 2 seems manifestly erroneous to me--the caller is trying to drop a lock it doesn't own. I think case 1 is also going to be a bug, but it technically depends on what the instance update saga ends up looking like. If we ignore the Failed state for a minute, and the saga guarantees the following, then it shouldn't be possible to drop a lock for a destroyed instance:

a. The saga only acquires the lock when an instance has an active VMM.
b. If the saga decides to retire an instance's VMMs and move the instance to Stopped, then it will drop the lock in the same atomic database operation.

If these rules hold, an instance can only be destroyed when the lock's not held, and so an instance should never get deleted out from under a lock owner.

The Failed state makes all this more complicated because Nexus will happily move an instance to Failed of its own volition, outside of anything the instance update functions might be doing, and it's possible to delete a Failed instance. We'll need to consider this carefully as we draft the instance update saga (and we should also make Nexus use the Failed state more judiciously, but that's for another set of PRs).

Reply:
Hmm, my thinking was just that:
  1. we (will) have a saga action that releases a lock
  2. saga actions must be idempotent
  3. therefore, releasing a lock must be idempotent

But, I didn't think *too* hard about it, just treating "saga actions must be idempotent" as a hard requirement. 

What I don't want, though, is a situation where a Nexus instance releases the lock but then crashes before it has recorded execution of that saga node, and then a subsequent Nexus resumes execution of that saga, attempts to release the lock again, and fails because it has already been unlocked, and then unwinds the whole saga, undoing a bunch of otherwise fine state transitions. Is that a valid concern?

Reply:
You're right, I missed the idempotency angle here. Without further refining this query I think that does require us to get an `Ok` value here. (The query seems to handle this case correctly: if someone presenting some lock ID gets a result of `Updated`, then repeats the query, the filter on line 711 should prevent the lock from being released again.)

Reply:
Okay, I _think_ the change from ceebdbb1abcd47f8ddf58638827a9f7206f48063 is a reasonable compromise between idempotency and "trying to unlock someone else's lock is Always Wrong And Bad". Now, `instance_updater_unlock` takes a generation number as well as a saga lock ID, and we return `Ok(false)` if the unlock query didn't do anything because the *generation number* has changed, but we return an error if the generation number is correct but the lock is held by a different saga. This way, if you unlock your lock twice, that doesn't error, but if you try to unlock someone *else*'s lock at the correct generation, you get an error.

What do you think?